### PR TITLE
feat: Allow saving with custom objectId

### PR DESCRIPTION
**Pull request by dplewis**
Closes: https://github.com/parse-community/Parse-SDK-JS/issues/1097
Ref: https://github.com/parse-community/parse-server/pull/6177

@JeromeDeLeon

### CODE REVIEW

**Review by mtrezza — APPROVED** [review]
LGTM!

File: src/ParseObject.js
Comment (JeromeDeLeon) [inline src/ParseObject.js:333]: Is there any purpose for this?

**Review by JeromeDeLeon — COMMENTED** [review]
LGTM! Just a question, if we were to merge this and update the Parse-Server, would it be allowed to use with save trigger?

File: src/ParseObject.js
Comment (dplewis) [reply src/ParseObject.js:333]: Nope

**Review by JeromeDeLeon — APPROVED** [review]

**Comment by dplewis** [comment]
@JeromeDeLeon The purpose of this PR is to allow the customId to get sent to the server. I wrote a quick test and the beforeSave trigger gets called but the behavior looks wrong. Might need a server update.

File: src/ParseObject.js
Comment (mstniy) [inline src/ParseObject.js:2387]: This forces all objects to have custom ids, if custom ids are allowed.
Is this intentional? This is not how the server treats allowCustomObjectId.

**Comment by breitman** [comment]
Hi, I am trying to set my own Id with this new feature. In my parse options I added the allowCustomObjectId to be true. Is there another thing i need to do with Parse JS SDK?

**Comment by mstniy** [comment]
Currently, one also needs to set allowCustomObjectId on the client side, which I find to be weird. It is used to decide between doing a POST and a PUT, in conjunction with updatedAt, which I find to be even weirder. Perhaps a later PR can modify the js sdk to have a create()/update() seperation instead, like the Dart sdk?

**Comment by mtrezza** [comment]
> Currently, one also needs to set allowCustomObjectId on the client side, which I find to be weird. It is used to decide between doing a POST and a PUT, in conjunction with updatedAt, which I find to be even weirder.

That indeed needs a closer look. Can you give more details / examples for that behavior?

**Comment by breitman** [comment]
> Currently, one also needs to set allowCustomObjectId on the client side, which I find to be weird. It is used to decide between doing a POST and a PUT, in conjunction with updatedAt, which I find to be even weirder. Perhaps a later PR can modify the js sdk to have a create()/update() seperation instead, like the Dart sdk?

So what is the notation to allow for custom IDs on the client side? Parse-server automatically creates the js sdk so i am confused as where i declare it on client side

**Comment by mstniy** [comment]
`Parse.allowCustomObjectId = true;` , as you can also see in the unit tests added by this PR. It can go anywhere, but it must be set by the time you call .save()
Another quirk is that this flag forces all objects to have a custom id, so it is really a `forceCustomObjectId` . I would love for this to change. In the meanwhile, do not forget to set it back to false after the operation involving custom object ids.

I am not sure what you mean by the server automatically creating the sdk.

**Comment by breitman** [comment]
> `Parse.allowCustomObjectId = true;` , as you can also see in the unit tests added by this PR. It can go anywhere, but it must be set by the time you call .save()
> Another quirk is that this flag forces all objects to have a custom id, so it is really a `forceCustomObjectId` . I would love for this to change. In the meanwhile, do not forget to set it back to false after the operation involving custom object ids.
> 
> I am not sure what you mean by the server automatically creating the sdk.

That isn't working. I added that line to the top of my cloud function, and i am still getting an object not found error, due to the custom id....

**Comment by mstniy** [comment]
Weird. Unless you are using an outdated version of the server, you might want to open an issue.
I can see the latest version on npm is 4.5.0, which was released before this PR was merged.

**Comment by mtrezza** [comment]
> In the meanwhile, do not forget to set it back to false after the operation involving custom object ids.

Are you intending to switch between custom and non-custom IDs during runtime? I'm not sure that should even be possible.

**Comment by mstniy** [comment]
> > Currently, one also needs to set allowCustomObjectId on the client side, which I find to be weird. It is used to decide between doing a POST and a PUT, in conjunction with updatedAt, which I find to be even weirder.
> 
> That indeed needs a closer look. Can you give more details / examples for what you described?
> 
> The difference between POST and PUT should whether the object already exists or is newly created, not how the object ID is set.

Please see [the code](https://github.com/parse-community/Parse-SDK-JS/blob/7512f533026431f7a085fb3f88d4b5c30430877c/src/ParseObject.js#L330-L337). Because the server doesn't support upsert (see parse-community/parse-server#7139), the SDK ends up having to guess whether to use POST (which fails if the object already exists) or PUT (which fails if the object does not already exist). The way it dos that is by checking if `createdAt` is set.

To see the way the Flutter SDK does it, see [its source](https://github.com/parse-community/Parse-SDK-Flutter/blob/aa6bed5df425ed1564261f7213177135bedb041a/packages/dart/lib/src/objects/parse_object.dart). It has a `.create()` that always does a `POST`, an`.update()` that always does a `PUT`, as well as a `.save()` that tries to guess.

**Comment by mstniy** [comment]
> > In the meanwhile, do not forget to set it back to false after the operation involving custom object ids.
> 
> Are you intending to switch between custom and non-custom IDs during runtime? I'm not sure such a fundamental change should be possible during run-time, or even at all for a deployment.

We want to _allow_ custom ids, but the changes in this PR forces us to either always use custom object ids, or to disallow them altogether. The way we currently work around this is by toggling the flag.

**Comment by lsmilek1** [comment]
In my application I found also a use case for the mixed environment of server generated and custom objectId ([ParseSwift SDK](https://github.com/parse-community/Parse-Swift/issues/220), [forum](https://community.parseplatform.org/t/both-custom-and-generated-objectid-in-one-project/1812)) and would like to implement *allow* instead of *force*.

From what I tested with forked ParseSwift SDK where I removed the guards that are checking that `objectId` is always set when `allowCustomObjectId = true` on both the server and the client, it works as intended:
- creating object with server generated `objectId` when `ParseObject.objectId = nil` and `ParseObject.createdAt = nil`
- creating object with custom `objectId` when `ParseObject.objectId = customId` and `ParseObject.createdAt = nil`
- updating object with any `objectId` when `ParseObject.objectId = any id` and `ParseObject.createdAt = any date`
- it throws error "Duplicate value" when trying to create object with custom id that already exists

So from this point of view there is no urgent need for separate `.create()` and `.update()` as long as it would be clearly documented that the `createdAt` is needed to decide whether the object should be updated or created - at least from client SDK, I am not 100% sure thing could break on server side JS SDK. But again, as it is already working and together with the documented need of `createdAt`, is there something blocking this to be implemented? In my use case if would safe few fields and indexes in the database together with extra fetches in the cloud code triggers.

**Comment by cbaker6** [comment]
> This forces all objects to have custom ids, if custom ids are allowed.
Is this intentional? This is not how the server treats allowCustomObjectId.

Though what @mstniy stated above is true. I don't think removing the checks that @dplewis added should be done as they make sure the proper items are needed when using a custom objectId. Instead, I recommend using a approach similar to the Swift SDK https://github.com/parse-community/Parse-Swift/pull/222#issue-984092514 in which a flag, `` is added and the developer wants to ignore the check to use a mixed environment, they can. The reasons for this is in the warnings added to the documentation:

**Comment by lsmilek1** [comment]
I see, that could be also a way to go. I am just still missing what is being protected exactly or at least it feels redundant, because from what I tested in ParseSwift SDK, it behaves correctly (assuming the `allowCustomObjectId = true` has to be set on the server). To summarise the behaviour I have in mind:

1) `allowCustomObjectId = false`
- `objectId = nil` --> save new object
- `objectId = generatedId` --> update object with `objectId == generatedId` or throw "does not exist"

2) `allowCustomObjectId = true`
- `objectId = nil`, `createdAt = nil` --> save new object with generated id
- `objectId = customId`, `createdAt = some date` --> save new object with custom id
- `objectId = someId`, `createdAt = some date` --> update object with `objectId == someId` or throw "does not exist"
- `objectId = someId`, `createdAt = nil` --> throw "duplicate value"

At least with ParseSwift SDK it work that way when I turned of the guards in forked SDK

File: src/ParseObject.js
Comment (bdevore17) [reply src/ParseObject.js:2387]: I second this. Just turned this on, and this does not work as expected.

Comment (cbaker6) [reply src/ParseObject.js:2387]: I recommend looking at https://github.com/parse-community/Parse-SDK-JS/pull/1309#issuecomment-909436697

**Pull request merged**
